resource_estimation/Error_Budget_Optimization/generate_data.py

```diff
@@ -0,0 +1,139 @@
+from __future__ import annotations
+
+import os
+from pathlib import Path
+
+import numpy as np
+from IPython.display import clear_output
+from qiskit import QuantumCircuit
+from qsharp.estimator import ErrorBudgetPartition, EstimatorParams, LogicalCounts
+from qsharp.interop.qiskit import estimate
+from tqdm import tqdm
+
+
+def find_optimized_budgets(total_budget, num_iterations, counts):
+    """
+    Randomly distributes the total error budget among logical, T-state, and rotation errors,
+    and estimates the physical resource requirements for each distribution. Tracks the distribution
+    that yields the lowest product of runtime and physical qubits.
+
+    Args:
+        total_budget: The total error budget to be distributed.
+        num_iterations: Number of random distributions to try.
+        counts: LogicalCounts object containing circuit logical counts.
+
+    Returns:
+        A tuple containing:
+            - List of optimal logical, T-state, and rotation budgets found.
+            - The best metric found (runtime * physical qubits).
+            - The default metric (runtime * physical qubits with default partition).
+    """
+    parameters = EstimatorParams()
+    default_parameters = EstimatorParams()
+    default_parameters.error_budget = total_budget
+
+    default_result = counts.estimate(default_parameters)
+    default_result["logicalCounts"]
+
+    default_physicalqubits = default_result["physicalCounts"]["physicalQubits"]
+    default_runtime = default_result["physicalCounts"]["runtime"]
+
+    running_metric = None
+
+    running_optimal_parameters = {}
+    logical_cache = []
+    t_cache = []
+    rotation_cache = []
+
+    for _i in range(num_iterations):
+        parameters.error_budget = ErrorBudgetPartition()
+        logical_budget_random = np.random.uniform(0, 1)
+        t_budget_random = np.random.uniform(0, 1)
+        rotation_budget_random = np.random.uniform(0, 1)
+        budget_sum = logical_budget_random + t_budget_random + rotation_budget_random
+
+        logical_budget = (logical_budget_random / budget_sum) * total_budget
+        t_budget = (t_budget_random / budget_sum) * total_budget
+        rotation_budget = (rotation_budget_random / budget_sum) * total_budget
+
+        parameters.error_budget.logical = logical_budget
+        parameters.error_budget.t_states = t_budget
+        parameters.error_budget.rotations = rotation_budget
+
+        result = counts.estimate(params=parameters)
+        default_result = counts.estimate()
+
+        physicalqubits = result["physicalCounts"]["physicalQubits"]
+        runtime = result["physicalCounts"]["runtime"]
+
+        default_metric = default_runtime * default_physicalqubits
+        current_metric = runtime * physicalqubits
+
+        if running_metric is None:
+            running_metric = current_metric
+            logical_cache = logical_budget
+            t_cache = t_budget
+            rotation_cache = rotation_budget
```

Comment:
Why not simply copy the `parameters` and keep track of that throughout instead of tracking in separate variables?
The function could also simply return the best set of parameters then.